wpcom-legacy-redirector.php

```diff
@@ -170,12 +171,112 @@ static function get_redirect_post_id( $url ) {
 
 		$redirect_post_id = $wpdb->get_var( $wpdb->prepare( "SELECT ID FROM $wpdb->posts WHERE post_type = %s AND post_name = %s LIMIT 1", self::POST_TYPE, $url_hash ) );
 
-		if ( ! $redirect_post_id )
+		if ( ! $redirect_post_id ) {
 			$redirect_post_id = 0;
+		}
 
 		return $redirect_post_id;
 	}
 
+	/**
+	 * Validate the redirect should work normally.
+	 *
+	 * @param array $redirect Redirect array.
+	 * @param array $post_types Array of publicly accessible post types.
+	 * @return bool|string True if the redirect passes validation. Array containing error notice if redirect fails validation.
+	 */
+	static function validate_redirect( $redirect, $post_types ) {
+
+		if ( 'url' === $redirect['destination_type'] ) {
+
+			if ( ! wp_validate_redirect( $redirect['to']['formatted'], false ) ) {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['formatted'],
+					'message'   => 'failed wp_validate_redirect()',
+				);
+			}
+
+		} else {
+
+			if ( ! $redirect['parent']['id'] > 0 ) {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['raw'],
+					'message'   => 'Attempting to redirect to a nonexistent post id.',
+				);
+			}
+
+			if (
+				'publish' !== $redirect['parent']['status']
+				&& 'attachment' !== $redirect['parent']['post_type']
+			) {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['raw'],
+					'message'   => 'Attempting to redirect to an unpublished post.',
+				);
+			}
+
+			if ( ! in_array( $redirect['parent']['post_type'], $post_types ) ) {
+				return array(
+					'id'        => $redirect['id'],
+					'from_url'  => $redirect['from']['path'],
+					'to_url'    => $redirect['to']['raw'],
+					'message'   => 'Attempting to redirect to a private post type: ' . $redirect['parent']['post_type'],
+				);
+			}
+		} // End if().
+
+		return true;
+	}
+
+	/**
+	 * Verify a redirect.
+	 *
+	 * @param array $redirect Redirect array.
+	 * @return bool|string True if the redirect passes validation. Array containing error notice if redirect fails validation.
+	 */
+	static function verify_redirect( $redirect ) {
+
+		if ( 3 == substr( $redirect['redirect']['status'], 0, 1 ) ) {
```

Comment:
This feels like an overly complicated way to verify the status code. We should either add an inline comment explaining what we're trying to do, or (preferably) just simplify to a greater than/less than check instead.